geo/src/algorithm/buffer.rs

```diff
@@ -0,0 +1,835 @@
+    fn clone_line_cap(&self) -> LineCap<BoolOpsCoord<T>, T> {
+        match &self.line_cap {
+            LineCap::Butt => LineCap::Butt,
+            LineCap::Round(angle) => LineCap::Round(*angle),
+            LineCap::Square => LineCap::Square,
+            LineCap::Custom(points) => LineCap::Custom(points.clone()),
+        }
+    }
+
+    fn default_join_angle() -> T {
+        // This is arbitrary, but the results seem to match pretty closely with JTS's own default.
+        T::from_f32(0.20).expect("valid float constant")
+    }
+
+    // Used by i_overlay for buffering (Multi)Polygons
+    fn outline_style(&self) -> OutlineStyle<T> {
+        OutlineStyle::new(self.distance).line_join(self.clone_line_join())
+    }
+
+    // Used by i_overlay for buffering (Multi)LineStrings
+    fn stroke_style(&self) -> StrokeStyle<BoolOpsCoord<T>, T> {
+        // "Buffer width" is like radius, whereas "stroke width" is more like diameter, so double the
+        // "stroke width" to reconcile semantics with "buffer width".
+        let two = T::one() + T::one();
+        StrokeStyle::new(self.distance * two)
+            .line_join(self.clone_line_join())
+            .end_cap(self.clone_line_cap())
+            .start_cap(self.clone_line_cap())
+    }
+}
+
+impl<F: BoolOpsNum + 'static> Buffer for Geometry<F> {
+    type Scalar = F;
+    crate::geometry_delegate_impl! {
+        fn buffer_with_style(&self, style: BufferStyle<Self::Scalar>) -> MultiPolygon<Self::Scalar>;
+    }
+}
+
+impl<F: BoolOpsNum + 'static> Buffer for Point<F> {
+    type Scalar = F;
+    fn buffer_with_style(&self, style: BufferStyle<Self::Scalar>) -> MultiPolygon<Self::Scalar> {
+        if style.distance <= F::zero() {
+            return MultiPolygon::new(vec![]);
+        }
+
+        match style.line_cap {
+            // Without length, i_overlay can't sensible reason about orientation of template points.
+            LineCap::Custom(_) => MultiPolygon::new(vec![]),
+            LineCap::Butt => MultiPolygon::new(vec![]),
+            LineCap::Square => {
+                let a = coord!(x: self.x() - style.distance, y: self.y() - style.distance);
+                let b = coord!(x: self.x() + style.distance, y: self.y() + style.distance);
+                MultiPolygon(vec![Rect::new(a, b).to_polygon()])
+            }
+            LineCap::Round(angle) => {
+                // approximate a circle
+                let num_segments = (2.0 * std::f64::consts::PI / angle.to_f64()).ceil() as usize;
+                let center = self.0;
+                let radius = style.distance;
+
+                let mut coords = Vec::with_capacity(num_segments + 1);
+                for i in 0..num_segments {
+                    let angle =
+                        F::from_f64(2.0 * std::f64::consts::PI / num_segments as f64 * i as f64)
+                            .expect("valid float constant");
+                    let x = center.x + radius * angle.cos();
```

Comment:
Huh, I kind of thought [sin_cos](https://doc.rust-lang.org/std/primitive.f64.html#method.sin_cos) would somehow be more efficient than the two individually, but apparently not

Reply:
It would be, but it's a very common compiler optimization.